File: src/NetFxCSharp/Controls/ADNullableValueSubControls/ADEditableValueForNullableValueControl.vb
Imports System.Windows.Forms

Friend Class ADEditableValueForNullableValueControl
    Inherits ADEditableValueForNullableValueControlTemplate(Of String)

    Protected WithEvents myTextBox As TextBox

    Sub New()
        MyBase.New()
        myTextBox = New TextBox
        myControl = myTextBox
        myControl.Name = "ADVariantTextControl" & myInstanceCounter
    End Sub

    Public Overrides Property Size() As System.Drawing.Size
        Get
            Return myControl.Size
        End Get
        Set(ByVal Value As System.Drawing.Size)
            myControl.Size = Value
        End Set
    End Property

    Public Overrides Function MeasureHeight() As Integer
        If DirectCast(myControl, TextBox).Multiline Then
            Return Me.Font.Height
        Else
            Return (myControl.Height) * -1
        End If
    End Function

    Public Overridable Property Multiline() As Boolean
        Get
            Return DirectCast(myControl, TextBox).Multiline
        End Get
        Set(ByVal Value As Boolean)
            DirectCast(myControl, TextBox).Multiline = Value
        End Set
    End Property

    Public Overrides Property Value() As Object
        Get
            Return DirectCast(myControl, TextBox).Text
        End Get
        Set(ByVal Value As Object)
            If Value Is Nothing Then
                DirectCast(myControl, TextBox).Text = ""
            Else
                DirectCast(myControl, TextBox).Text = Value.ToString()
            End If
        End Set
    End Property

    Public Overridable Property ScrollBars() As ScrollBars
        Get
            Return DirectCast(myControl, TextBox).ScrollBars
        End Get
        Set(ByVal Value As ScrollBars)
            DirectCast(myControl, TextBox).ScrollBars = Value
        End Set
    End Property

    Public Overridable Property MaxLength() As Integer
        Get
            Return DirectCast(myControl, TextBox).MaxLength
        End Get
        Set(ByVal Value As Integer)
            DirectCast(myControl, TextBox).MaxLength = Value
        End Set
    End Property

    Public Overridable Property [Readonly]() As Boolean
        Get
            Return DirectCast(myControl, TextBox).ReadOnly
        End Get
        Set(ByVal value As Boolean)
            DirectCast(myControl, TextBox).ReadOnly = value
        End Set
    End Property

    Private Sub myTextBox_TextChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles myTextBox.TextChanged
        OnValueChanged(e)
    End Sub

    Private Sub myTextBox_Validated(ByVal sender As Object, ByVal e As System.EventArgs) Handles myTextBox.Validated
        OnValidated(e)
    End Sub

    Private Sub myTextBox_Validating(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles myTextBox.Validating
        OnValidating(e)
    End Sub
End Class

File: src/NetFxCSharp/Controls/ADNullableValueSubControls/ADEditableValueForNullableValueControl.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Xml.Linq;

namespace ActiveDev.Controls
{
    internal class ADEditableValueForNullableValueControl : ADEditableValueForNullableValueControlTemplate<string>
    {
        private TextBox _myTextBox;
        protected TextBox myTextBox
        {
            get
            {
                return _myTextBox;
            }

            set
            {
                if (_myTextBox != null)
                {
                    _myTextBox.TextChanged -= myTextBox_TextChanged;
                    _myTextBox.Validated -= myTextBox_Validated;
                    _myTextBox.Validating -= myTextBox_Validating;
                }

                _myTextBox = value;
                if (_myTextBox != null)
                {
                    _myTextBox.TextChanged += myTextBox_TextChanged;
                    _myTextBox.Validated += myTextBox_Validated;
                    _myTextBox.Validating += myTextBox_Validating;
                }
            }
        }

        public ADEditableValueForNullableValueControl() : base()
        {
            myTextBox = new TextBox();
            myControl = myTextBox;
            myControl.Name = "ADVariantTextControl" + myInstanceCounter;
        }

        public override System.Drawing.Size Size
        {
            get
            {
                return myControl.Size;
            }

            set
            {
                System.Drawing.Size Value = value;
                myControl.Size = Value;
            }
        }

        public override int MeasureHeight()
        {
            if (((TextBox)myControl).Multiline)
            {
                return this.Font.Height;
            }
            else
            {
                return (myControl.Height) * -1;
            }
        }

        public virtual bool Multiline
        {
            get
            {
                return ((TextBox)myControl).Multiline;
            }

            set
            {
                bool Value = value;
                ((TextBox)myControl).Multiline = Value;
            }
        }

        public override object Value
        {
            get
            {
                return ((TextBox)myControl).Text;
            }

            set
            {
                object Value = value;
                if (Value == null)
                {
                    ((TextBox)myControl).Text = "";
                }
                else
                {
                    ((TextBox)myControl).Text = Value.ToString();
                }
            }
        }

        public virtual ScrollBars ScrollBars
        {
            get
            {
                return ((TextBox)myControl).ScrollBars;
            }

            set
            {
                ScrollBars Value = value;
                ((TextBox)myControl).ScrollBars = Value;
            }
        }

        public virtual int MaxLength
        {
            get
            {
                return ((TextBox)myControl).MaxLength;
            }

            set
            {
                int Value = value;
                ((TextBox)myControl).MaxLength = Value;
            }
        }

        public virtual bool Readonly
        {
            get
            {
                return ((TextBox)myControl).ReadOnly;
            }

            set
            {
                ((TextBox)myControl).ReadOnly = value;
            }
        }

        private void myTextBox_TextChanged(object sender, System.EventArgs e)
        {
            OnValueChanged(e);
        }

        private void myTextBox_Validated(object sender, System.EventArgs e)
        {
            OnValidated(e);
        }

        private void myTextBox_Validating(object sender, System.ComponentModel.CancelEventArgs e)
        {
            OnValidating(e);
        }
    }
}